File: lib/models/workout.dart
/// Maps the backend's `WorkoutOut` schema.
class Workout {
  const Workout({
    required this.id,
    required this.exercise,
    required this.totalReps,
    required this.durationSeconds,
    required this.accuracyScore,
    required this.startedAt,
    required this.endedAt,
    required this.createdAt,
  });

  final int id;
  final String exercise;
  final int totalReps;
  final double? durationSeconds;
  final double? accuracyScore;
  final DateTime startedAt;
  final DateTime? endedAt;
  final DateTime createdAt;

  factory Workout.fromJson(Map<String, dynamic> json) => Workout(
        id: json['id'] as int,
        exercise: json['exercise'] as String,
        totalReps: json['total_reps'] as int,
        durationSeconds: (json['duration_seconds'] as num?)?.toDouble(),
        accuracyScore: (json['accuracy_score'] as num?)?.toDouble(),
        startedAt: DateTime.parse(json['started_at'] as String),
        endedAt: json['ended_at'] == null
            ? null
            : DateTime.parse(json['ended_at'] as String),
        createdAt: DateTime.parse(json['created_at'] as String),
      );
}

File: lib/backend/app/models/workout.py
"""Model bảng workouts — lịch sử buổi tập."""

from datetime import datetime, timezone
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base

if TYPE_CHECKING:
    from app.models.user import User


class Workout(Base):
    __tablename__ = "workouts"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("Users.UserId"), nullable=False, index=True)
    exercise: Mapped[str] = mapped_column(String(100), nullable=False)
    total_reps: Mapped[int] = mapped_column(Integer, default=0)
    duration_seconds: Mapped[float | None] = mapped_column(Float, nullable=True)
    accuracy_score: Mapped[float | None] = mapped_column(Float, nullable=True)
    started_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    ended_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    user: Mapped["User"] = relationship("User", back_populates="workouts")
